src/stream/src/common/log_store/mod.rs

```diff
@@ -0,0 +1,270 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::Debug;
+use std::future::Future;
+
+use risingwave_common::array::StreamChunk;
+use risingwave_common::util::epoch::INVALID_EPOCH;
+use tokio::sync::mpsc::{
+    channel, unbounded_channel, Receiver, Sender, UnboundedReceiver, UnboundedSender,
+};
+use tokio::sync::oneshot;
+
+use crate::common::log_store::LogReaderEpochProgress::{AwaitingTruncate, Consuming};
+
+#[derive(thiserror::Error, Debug)]
+pub enum LogStoreError {
+    #[error("EndOfLogStream")]
+    EndOfLogStream,
+}
+
+pub type LogStoreResult<T> = Result<T, LogStoreError>;
+
+#[derive(Debug)]
+pub enum LogStoreReadItem {
+    StreamChunk(StreamChunk),
+    Barrier {
+        next_epoch: u64,
+        is_checkpoint: bool,
+    },
+}
+
+pub trait LogWriter {
```

Comment:
What is the granularity of LogWriter and LogReader? 
1. If it is per actor, internally the consumption progress should be track on a per vnode basis
2. If it is per vnode, each sink executor will have multiple LogWriter/LogReaders and message needs to be dispatched correctly

Not sure whether the above choices will affect the LogWriter and LogReader interface. Just a reminder. We can leave it to future PR.

Reply:
The design is per actor. I have added a method `update_vnode_mapping` to handle potential need for scale.